DSCResources/cChocoPackageInstall/cChocoPackageInstall.psm1

```diff
@@ -389,41 +383,81 @@ Function Upgrade-Package {
     if ($cParams) {
         $chocoupgradeparams += " $cParams"
     }
-    $cmd = "choco upgrade $pName $chocoupgradeparams"
+    $cmd = "upgrade $pName $chocoupgradeparams"
     Write-Verbose -Message "Upgrade command: '$cmd'"
 
     if (-not (IsPackageInstalled -pName $pName))
     {
         throw "$pName is not installed, you cannot upgrade"
     }
 
-    $packageUpgradeOuput = Invoke-Expression -Command $cmd
+    $packageUpgradeOuput = Invoke-Chocolatey $cmd
     $packageUpgradeOuput | ForEach-Object { Write-Verbose -Message $_ }
+}
 
-    # Clear Package Cache
-    Get-ChocoInstalledPackage 'Purge'
+function Get-ChocoInstalledPackage {
+    Return (choco list -lo -r | ConvertFrom-Csv -Header 'Name', 'Version' -Delimiter "|")
 }
 
-function Get-ChocoInstalledPackage ($action) {
-    $ChocoInstallLP = Join-Path -Path $env:ChocolateyInstall -ChildPath 'cache'
-    if ( -not (Test-Path $ChocoInstallLP)){
-        New-Item -Name 'cache' -Path $env:ChocolateyInstall -ItemType Directory | Out-Null
-    }
-    $ChocoInstallList = Join-Path -Path $ChocoInstallLP -ChildPath 'ChocoInstalled.xml'
 
-    if ($action -eq 'Purge') {
-        Remove-Item $ChocoInstallList -Force
-        $res = $true
-    } else {
-        $PackageCacheSec = (Get-Date).AddSeconds('-60')
-        if ( $PackageCacheSec -lt (Get-Item $ChocoInstallList -ErrorAction SilentlyContinue).LastWriteTime ) {
-                $res = Import-Clixml $ChocoInstallList
-        } else {
-            choco list -lo -r | ConvertFrom-Csv -Header 'Name', 'Version' -Delimiter "|" -OutVariable res | Export-Clixml -Path $ChocoInstallList
-        }
-    }
+<#
+.Synopsis
+   Run chocolatey executable and throws error on failure
+.DESCRIPTION
+   Run chocolatey executable and throws error on failure
+.EXAMPLE
+   Invoke-Chocolatey "list -lo"
+.EXAMPLE
+   Invoke-Chocolatey -arguments "list -lo"
+#>
+function Invoke-Chocolatey
+{
+    [CmdletBinding()]
+    Param
+    (
+        # chocolatey arguments."
+        [Parameter(Position=0)]
+        [string]$arguments
+    )
 
-    Return $res
+    [int[]]$validExitCodes =  $(
+                0,    #most widely used success exit code
+                1605, #(MSI uninstall) - the product is not found, could have already been uninstalled
+                1614, #(MSI uninstall) - the product is uninstalled
+                1641, #(MSI) - restart initiated
+                3010  #(MSI, InnoSetup can be passed to provide this) - restart required
+            )
+    Write-Verbose -Message "command: 'choco $arguments'" 
+    
+    $pinfo = New-Object System.Diagnostics.ProcessStartInfo
+    $pinfo.FileName = "choco"
+    $pinfo.RedirectStandardError = $true
+    $pinfo.RedirectStandardOutput = $true
+    $pinfo.UseShellExecute = $false
+    $pinfo.Arguments = "$arguments"
+
+    $p = New-Object System.Diagnostics.Process
+    $p.StartInfo = $pinfo
+    $p.Start() | Out-Null
+
+    $output = $p.StandardOutput.ReadToEnd()
+    $p.WaitForExit()

```

Comment:
By default choco.exe will timeout after 45 minutes. It can be changed in the config but I'm wondering if we should kill it before then. Thoughts?

Reply:
45 minutes is a long time, but i can imagin that in some instances an install of a package takes longer than that. For instance a full install of visual studio on a slow internet connection.